support/ebpf/ruby_tracer.ebpf.c

```diff
@@ -17,18 +18,280 @@ struct ruby_procs_t {
 // The number of Ruby frames to unwind per frame-unwinding eBPF program. If
 // we start running out of instructions in the walk_ruby_stack program, one
 // option is to adjust this number downwards.
-#define FRAMES_PER_WALK_RUBY_STACK 27
+// NOTE the maximum size stack is this times 33
+#define FRAMES_PER_WALK_RUBY_STACK 32
+// When resolving a CME, we need to traverse environment pointers until we
+// find IMEMO_MENT. Since we can't do a while loop, we have to bound this
+// the max encountered in experimentation on a production rails app is 6.
+// This increases insn for the kernel verifier all code in the ep check "loop"
+// is M*N for instruction checks, so be extra sensitive about additions there.
+// If we get ERR_RUBY_READ_CME_MAX_EP regularly, we may need to raise it.
+#define MAX_EP_CHECKS              6
+
+// Constants related to reading a method entry
+#define VM_METHOD_TYPE_ISEQ 0
+#define VM_ENV_FLAG_LOCAL   0x2
+#define RUBY_FL_USHIFT      12
+#define IMEMO_MASK          0x0f
+#define IMEMO_SVAR          2
+#define IMEMO_MENT          6
+
+// https://github.com/ruby/ruby/blob/v3_4_5/vm_core.h#L1380-L1385
+#define VM_FRAME_MAGIC_MASK  0x7fff0001
+#define VM_FRAME_MAGIC_CFUNC 0x55550001
+
+// Save on read ops by reading the whole control frame struct
+// as technically this reads too much memory
+typedef struct rb_control_frame_struct {
+  const void *pc;         // cfp[0]
+  void *sp;               // cfp[1]
+  const void *iseq;       // cfp[2]
+  void *self;             // cfp[3] / block[0]
+  const void *ep;         // cfp[4] / block[1]
+  const void *block_code; // cfp[5] / block[2] -- iseq, ifunc, or forwarded block handler
+  void *jit_return;       // cfp[6] -- return address for JIT code
+} rb_control_frame_t;
```

Comment:
Minor: I wonder if it's worth introducing a typedef for `VALUE`, to keep the closer alignment with the types used on the Ruby VM (rather than replacing them with a `void *`)